File: python/imcombiners/_offset.py
"""Offset/padding helpers."""

from __future__ import annotations

from collections.abc import Iterable
from typing import TYPE_CHECKING

import numpy as np
from astro_ndslice.offset import offseted_shape, offsets2slice

if TYPE_CHECKING:
    from numpy.typing import ArrayLike, NDArray

_FLOAT_DTYPES = (np.float32, np.float64)
_PROMOTE_TO_FLOAT32_DTYPES = (np.uint8, np.uint16, np.int16)
_PROMOTE_TO_FLOAT64_DTYPES = (np.int32,)


def _normalize_offset_image(image: ArrayLike) -> NDArray:
    """Validate one N-D array for offset padding."""
    arr = np.asarray(image)
    if arr.ndim < 1:
        raise ValueError(
            f"images must have at least 1 dimension; got shape {arr.shape}"
        )
    if 0 in arr.shape:
        raise ValueError(f"images must have non-zero axes; got shape {arr.shape}")
    if arr.dtype in _PROMOTE_TO_FLOAT32_DTYPES:
        return np.ascontiguousarray(arr, dtype=np.float32)
    if arr.dtype in _PROMOTE_TO_FLOAT64_DTYPES:
        return np.ascontiguousarray(arr, dtype=np.float64)
    if arr.dtype not in _FLOAT_DTYPES:
        raise TypeError(
            "images must be uint8, uint16, int16, int32, float32, "
            f"or float64; got {arr.dtype}"
        )
    return np.ascontiguousarray(arr)


def _normalize_offset_images(
    images: Iterable[ArrayLike],
) -> list[NDArray]:
    """Validate offset images and promote them to one floating dtype."""
    arrays = [_normalize_offset_image(image) for image in images]
    if not arrays:
        raise ValueError("images must contain at least one array")
    ndim = arrays[0].ndim
    for arr in arrays:
        if arr.ndim != ndim:
            raise ValueError("images must all have the same number of dimensions")
    dtype = np.result_type(*(arr.dtype for arr in arrays))
    if dtype not in (np.dtype(np.float32), np.dtype(np.float64)):
        dtype = np.dtype(np.float64)
    return [np.ascontiguousarray(arr, dtype=dtype) for arr in arrays]


def place_into_padded(
    images: Iterable[ArrayLike],
    offsets: ArrayLike,
    *,
    fill: float = np.nan,
    validate: bool = True,
) -> NDArray:
    """Place N-D arrays into a padded stack using integer offsets.

    Offsets are in Python/NumPy axis order. For 2-D images, `(dy, dx)` means
    `(row offset, column offset)`: positive `dy` places an image at a larger
    row index, and positive `dx` places it at a larger column index. Raw
    offsets are normalized by subtracting the per-axis minimum offset, matching
    `astro-ndslice` outer-shape convention for offsets already in Python axis
    order.

    Parameters
    ----------
    images : sequence of array-like, each shape (*spatial)
        Input arrays. Arrays may have different shapes, but must all have the
        same number of dimensions. Accepted dtypes are
        `uint8`, `uint16`, `int16`, `int32`, `float32`, and `float64`; integer
        images are promoted to a floating workspace. Other dtypes, including
        `int64` and `float128`, are not silently cast. Cast unsupported arrays
        explicitly before padding.
    offsets : array-like, shape (N, ndim)
        Integer offsets in NumPy axis order. `N` must match the number of
        arrays, and `ndim` must match each input array.
    fill : float, optional
        Value assigned to uncovered pixels in the padded stack. The default is
        `NaN`, which makes the combine kernels ignore uncovered regions.
    validate : bool, optional
        If `True`, validate image dimensionality, dtype, offset shape, and
        offset dtype. If `False`, callers must provide same-rank contiguous
        arrays with compatible dtypes and an integer offset array of shape
        `(N, ndim)`.

    Returns
    -------
    padded : ndarray, shape (N, *outer_shape)
        Stack containing all arrays placed into the common padded frame.
    """
    if validate:
        images = _normalize_offset_images(images)
        offsets = np.asarray(offsets)
        if not np.issubdtype(offsets.dtype, np.integer):
            raise TypeError(f"offsets must have integer dtype; got {offsets.dtype}")
        offsets = np.ascontiguousarray(offsets, dtype=np.int64)
    else:
        images = list(images)
        offsets = np.ascontiguousarray(offsets, dtype=np.int64)
    ndim = images[0].ndim
    if offsets.shape != (len(images), ndim):
        raise ValueError(
            f"offsets shape must be ({len(images)}, {ndim}); got {offsets.shape}"
        )

    shapes = np.array([image.shape for image in images], dtype=np.int64)
    int64_max = np.iinfo(np.int64).max
    intp_max = np.iinfo(np.intp).max
    min_offsets = offsets.min(axis=0)
    for image_shape, offset in zip(shapes, offsets, strict=True):
        for axis, (axis_len, raw_offset, min_offset) in enumerate(
            zip(image_shape, offset, min_offsets, strict=True)
        ):
            label = "y" if axis == 0 else "x" if axis == 1 else f"axis {axis}"
            start = int(raw_offset) - int(min_offset)
            if start > int64_max:
                raise ValueError(f"normalized {label} offset overflow")
            if start > intp_max:
                raise ValueError(f"normalized {label} offset does not fit usize")
            if start + int(axis_len) > intp_max:
                raise ValueError(f"padded output {label} overflow")

    _, outer_shape = offseted_shape(
        shapes,
        offsets,
        method="outer",
        offset_order_xyz=False,
        intify_offsets=True,
        pythonize_offsets=True,
    )
    slices = offsets2slice(
        shapes,
        offsets,
        method="outer",
        shape_order_xyz=False,
        offset_order_xyz=False,
        outer_for_stack=False,
        fits_convention=False,
    )
    padded = np.full((len(images), *outer_shape), fill, dtype=images[0].dtype)
    for image_index, (image, image_slices) in enumerate(
        zip(images, slices, strict=True)
    ):
        padded[(image_index, *image_slices)] = image
    return padded

File: src/kernel/offset.rs
//! Offset/padding helpers for array stacks.

use ndarray::{ArrayD, ArrayViewD, Dimension, IxDyn};

use super::utils::Float;

#[inline]
fn axis_label(axis: usize) -> String {
    match axis {
        0 => "y".to_string(),
        1 => "x".to_string(),
        _ => format!("axis {axis}"),
    }
}

fn normalized_offsets_and_shape(
    shapes: &[Vec<usize>],
    offsets: &[Vec<i64>],
) -> Result<(Vec<Vec<usize>>, Vec<usize>), String> {
    if shapes.is_empty() {
        return Err("images must contain at least one array".to_string());
    }
    if shapes.len() != offsets.len() {
        return Err(format!(
            "offsets length {} does not match image count {}",
            offsets.len(),
            shapes.len()
        ));
    }
    let ndim = shapes[0].len();
    if ndim == 0 {
        return Err("images must have at least 1 dimension".to_string());
    }
    for shape in shapes {
        if shape.len() != ndim {
            return Err("images must all have the same number of dimensions".to_string());
        }
    }
    for offset in offsets {
        if offset.len() != ndim {
            return Err(format!(
                "offsets shape must be ({}, {}); got width {}",
                shapes.len(),
                ndim,
                offset.len()
            ));
        }
    }

    let mins: Vec<i64> = (0..ndim)
        .map(|axis| offsets.iter().map(|offset| offset[axis]).min().unwrap())
        .collect();
    let mut starts = Vec::with_capacity(offsets.len());
    let mut out_shape = vec![0usize; ndim];

    for (shape, offset) in shapes.iter().zip(offsets.iter()) {
        let mut start = Vec::with_capacity(ndim);
        for axis in 0..ndim {
            let label = axis_label(axis);
            let normalized = offset[axis]
                .checked_sub(mins[axis])
                .ok_or_else(|| format!("normalized {label} offset overflow"))?;
            let pos = usize::try_from(normalized)
                .map_err(|_| format!("normalized {label} offset does not fit usize"))?;
            let stop = pos
                .checked_add(shape[axis])
                .ok_or_else(|| format!("padded output {label} overflow"))?;
            out_shape[axis] = out_shape[axis].max(stop);
            start.push(pos);
        }
        starts.push(start);
    }

    Ok((starts, out_shape))
}

pub fn place_into_padded<T: Float>(
    images: &[ArrayViewD<'_, T>],
    offsets: &[Vec<i64>],
    fill: T,
) -> Result<ArrayD<T>, String> {
    let shapes: Vec<Vec<usize>> = images.iter().map(|image| image.shape().to_vec()).collect();
    let (starts, outer_shape) = normalized_offsets_and_shape(&shapes, offsets)?;
    let mut out_shape = Vec::with_capacity(outer_shape.len() + 1);
    out_shape.push(images.len());
    out_shape.extend(outer_shape);
    let mut out = ArrayD::<T>::from_elem(IxDyn(&out_shape), fill);

    for (k, image) in images.iter().enumerate() {
        let start = &starts[k];
        for (idx, value) in image.indexed_iter() {
            let mut out_idx = Vec::with_capacity(idx.ndim() + 1);
            out_idx.push(k);
            for axis in 0..idx.ndim() {
                out_idx.push(start[axis] + idx[axis]);
            }
            out[IxDyn(&out_idx)] = *value;
        }
    }

    Ok(out)
}

#[cfg(test)]
mod tests {
    use ndarray::array;

    use super::place_into_padded;

    #[test]
    fn normalizes_negative_offsets_and_computes_outer_shape() {
        let img0 = array![[1.0_f32, 2.0], [3.0, 4.0]];
        let img1 = array![[10.0_f32, 11.0, 12.0]];
        let images = vec![img0.view().into_dyn(), img1.view().into_dyn()];
        let offsets = vec![vec![2_i64, -1_i64], vec![-1_i64, 1_i64]];

        let out = place_into_padded(&images, &offsets, f32::NAN).unwrap();

        assert_eq!(out.shape(), &[2, 5, 5]);
        assert_eq!(out[[0, 3, 0]], 1.0);
        assert_eq!(out[[0, 4, 1]], 4.0);
        assert_eq!(out[[1, 0, 2]], 10.0);
        assert_eq!(out[[1, 0, 4]], 12.0);
        assert!(out[[0, 0, 0]].is_nan());
    }
}
